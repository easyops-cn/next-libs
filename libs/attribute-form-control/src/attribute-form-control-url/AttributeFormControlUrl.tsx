import { FormControl } from "../model-attribute-form-control/ModelAttributeFormControl";
import { Input } from "antd";
import React, { Component, ReactNode } from "react";
import "./AttributeFormControlUrl.module.css";

export interface AttributeFormControlUrlProps extends Partial<FormControl> {
  value: string;
  onChange: (value: string) => void;
}

export class AttributeFormControlUrl extends Component<
  AttributeFormControlUrlProps
> {
  private url: string;
  private title: string;

  constructor(props: AttributeFormControlUrlProps) {
    super(props);

    const { title, url } = AttributeFormControlUrl.breakDownValue(props.value);
    this.title = title;
    this.url = url;
  }

  static breakDownValue(value: string): { title: string; url: string } {
    const pattern = /^\[(.*)\]\((.*)\)$/;
    const arr = value && value.match(pattern);

    if (arr) {
      return {
        title: arr[1],
        url: AttributeFormControlUrl.getFormattedUrl(arr[2])
      };
    }

    return {
      title: value,
      url: value
    };
  }

  static getFormattedUrl(url: string): string {
    if (url) {
      const pattern = /^(http|https):\/\/.*$/;

      return pattern.test(url) ? url : "http://" + url;
    }
    return "";
  }

  preSetValidation = () => {
    const url = AttributeFormControlUrl.getFormattedUrl(this.url);

    if (this.url !== url) {
      this.url = url;
      this.composeUrl();
    }
  };

  composeUrl(): void {
    const { url, title } = this;
    const value = url || title ? `[${title}](${url})` : "";

    this.props.onChange(value);
  }

  handleUrlChange = (url: string) => {
    this.url = url;
    this.composeUrl();
  };

  handleTitleChange = (title: string) => {
    this.title = title;
    this.composeUrl();
  };

  render(): ReactNode {
    const { readOnly, placeholder } = this.props;
    const { title, url } = this;

    return (
      <div>
        <div className="singleContainer">
          <label className="presetTitle">链接：</label>
          <Input
            readOnly={readOnly}
            value={url}
            type="url"
            onChange={e => this.handleUrlChange(e.target.value)}
            onBlur={this.preSetValidation}
            placeholder={placeholder || "请输入链接"}
          />
        </div>

        <div className="singleContainer">
          <label className="presetTitle">标题：</label>
          <Input
            defaultValue={title}
            type="text"
            onChange={e => this.handleTitleChange(e.target.value)}
            onBlur={this.preSetValidation}
            placeholder="选填，请输入显示的标题"
          />
        </div>
      </div>
    );
  }
}
