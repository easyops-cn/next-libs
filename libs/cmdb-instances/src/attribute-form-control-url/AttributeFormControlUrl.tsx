import { FormControl } from "../model-attribute-form-control/ModelAttributeFormControl";
import { Input } from "antd";
import i18n from "i18next";
import { NS_LIBS_CMDB_INSTANCES, K } from "../i18n/constants";
import React, { Component, ReactNode } from "react";
import "./AttributeFormControlUrl.module.css";

export interface AttributeFormControlUrlProps extends Partial<FormControl> {
  value: string;
  onChange: (value: string) => void;
}

export class AttributeFormControlUrl extends Component<AttributeFormControlUrlProps> {
  private url: string;
  private title: string;

  constructor(props: AttributeFormControlUrlProps) {
    super(props);

    const { title, url } = AttributeFormControlUrl.breakDownValue(props.value);
    this.title = title;
    this.url = url;
  }

  static breakDownValue(value: string): { title: string; url: string } {
    const pattern = /^\[(.*)\]\((.*)\)$/;
    const arr = value && value.match(pattern);
    if (arr) {
      return {
        title: arr[1],
        url: AttributeFormControlUrl.getFormattedUrl(arr[2]),
      };
    }

    return {
      title: value,
      url: value,
    };
  }

  static getFormattedUrl(url: string): string {
    /* istanbul ignore if  */
    if (url) {
      const pattern = /^(http|https):\/\/.*$/;
      // return pattern.test(url) ? url : "http://" + url;
      const urlReg = /^((http|https):\/\/)?(([A-Za-z0-9]+-[A-Za-z0-9]+|[A-Za-z0-9]+)\.)+([A-Za-z]+)[/\?\:]?.*|^((2[0-4]\d|25[0-5]|[01]?\d\d?)\.){3}(2[0-4]\d|25[0-5]|[01]?\d\d?)$/;
      const reg = /^\//
      const origin = window.location.origin;
      return pattern.test(url) ? url : urlReg.test(url) ? url : `${origin}${reg.test(url) ? url : '/' + url} `;
    }
    return "";
  }

  // preSetValidation = () => {
  //   const url = AttributeFormControlUrl.getFormattedUrl(this.url);

  //   if (this.url !== url) {
  //     this.url = url;
  //     this.composeUrl();
  //   }
  // };

  /* istanbul ignore next */
  composeUrl(): void {
    const title = this.title ?? "";
    const url = AttributeFormControlUrl.getFormattedUrl(this.url);
    const value = url || title ? `[${title}](${url})` : "";
    this.props.onChange(value);
  }

  handleUrlChange = (url: string) => {
    this.url = url;
    this.composeUrl();
  };

  handleTitleChange = (title: string) => {
    this.title = title;
    this.composeUrl();
  };

  render(): ReactNode {
    const { readOnly, placeholder } = this.props;
    const { title, url } = this;

    return (
      <div>
        <div className="singleContainer">
          <label className="presetTitle">
            {i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.LINK}`)}
          </label>
          <Input
            readOnly={readOnly}
            value={url}
            type="url"
            onChange={(e) => this.handleUrlChange(e.target.value)}
            // onBlur={this.preSetValidation}
            placeholder={
              placeholder ||
              i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.LINK_PLACEHOLDER}`)
            }
          />
        </div>

        <div className="singleContainer">
          <label className="presetTitle">
            {i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.TITLE}`)}
          </label>
          <Input
            defaultValue={title}
            type="text"
            onChange={(e) => this.handleTitleChange(e.target.value)}
            // onBlur={this.preSetValidation}
            placeholder={i18n.t(
              `${NS_LIBS_CMDB_INSTANCES}:${K.TITLE_PLACEHOLDER}`
            )}
          />
        </div>
      </div>
    );
  }
}
