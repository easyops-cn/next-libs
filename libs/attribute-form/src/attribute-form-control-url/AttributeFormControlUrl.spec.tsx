import React from "react";
import { mount, ReactWrapper } from "enzyme";
import { AttributeFormControlUrl } from "./AttributeFormControlUrl";

describe("AttributeFormControlUrl", () => {
  let wrapper: ReactWrapper<
    AttributeFormControlUrl["props"],
    AttributeFormControlUrl["state"],
    React.Component<{}, {}, any>
  >;
  let instance: React.Component<{}, {}, any>;
  const props = {
    value: "[百度](http://www.baidu.com)",
    onChange: jest.fn(),
    readOnly: true
  };

  beforeEach(() => {
    wrapper = mount(<AttributeFormControlUrl {...props} />);
    instance = wrapper.instance();
  });

  afterEach(() => {
    props.onChange.mockClear();
  });

  it("should change url value", () => {
    const urlEle = wrapper.find("input").at(0);

    urlEle.simulate("change", {
      target: {
        value: "www.google.com"
      }
    });

    expect(props.onChange).toHaveBeenCalledWith("[百度](www.google.com)");

    urlEle.simulate("blur");
    expect(props.onChange).toHaveBeenCalledWith(
      "[百度](http://www.google.com)"
    );
  });

  it("should change title value", () => {
    const titleEle = wrapper.find("input").at(1);

    titleEle.simulate("change", {
      target: {
        value: "google"
      }
    });

    expect(props.onChange).toHaveBeenCalledWith(
      "[google](http://www.baidu.com)"
    );

    titleEle.simulate("blur");
    expect(props.onChange).toHaveBeenCalledTimes(1);
  });

  it("value should be null when props.value to equal null", () => {
    const newProps = Object.assign({}, props, {
      value: null
    });
    const wrapper = mount(<AttributeFormControlUrl {...newProps} />);

    const [urlEle, titleEle] = wrapper.find("input").map(node => node);
    expect(urlEle.props().value).toBeFalsy();
    expect(titleEle.props().value).toBeFalsy();
  });

  it("should work", () => {
    const { title, url } = AttributeFormControlUrl.breakDownValue(props.value);
    const [urlEle, titleEle] = wrapper.find("input").map(node => node);

    expect(urlEle.exists() && titleEle.exists()).toBeTruthy();
    expect(urlEle.props().value).toEqual(url);
    expect(urlEle.props().placeholder).toBe("请输入链接");
    expect(urlEle.props().type).toBe("url");
    expect(urlEle.props().readOnly).toBeTruthy();

    expect(titleEle.props().value).toEqual(title);
    expect(titleEle.props().placeholder).toBe("选填，请输入显示的标题");
    expect(titleEle.props().type).toBe("text");

    expect(wrapper).toMatchSnapshot();
  });
});
