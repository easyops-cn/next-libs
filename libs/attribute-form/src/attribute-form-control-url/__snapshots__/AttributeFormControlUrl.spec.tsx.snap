// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeFormControlUrl should work 1`] = `
<AttributeFormControlUrl
  onChange={[MockFunction]}
  readOnly={true}
  value="[百度](http://www.baidu.com)"
>
  <div>
    <div
      className="singleContainer"
    >
      <label
        className="presetTitle"
      >
        链接：
      </label>
      <Input
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="请输入链接"
        readOnly={true}
        type="url"
        value="http://www.baidu.com"
      >
        <input
          className="ant-input"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="请输入链接"
          readOnly={true}
          type="url"
          value="http://www.baidu.com"
        />
      </Input>
    </div>
    <div
      className="singleContainer"
    >
      <label
        className="presetTitle"
      >
        标题：
      </label>
      <Input
        defaultValue="百度"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="选填，请输入显示的标题"
        type="text"
      >
        <input
          className="ant-input"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="选填，请输入显示的标题"
          type="text"
          value="百度"
        />
      </Input>
    </div>
  </div>
</AttributeFormControlUrl>
`;
